detection_rules/rule_validators.py

```diff
@@ -81,34 +97,44 @@ def validate(self, data: 'QueryRuleData', meta: RuleMeta) -> None:
             return
 
         indexes = data.index or []
-        beats_version = meta.beats_version or beats.get_max_version()
-        ecs_versions = meta.ecs_versions or [ecs.get_max_version()]
+        stack_versions = get_stack_schemas(meta.min_stack_version)
+
+        if not stack_versions:
+            # add latest
+            stack_versions = {CURRENT_PACKAGE: {'beats': beats.get_max_version(), 'ecs': ecs.get_max_version()}}
+
+        if meta.query_validate_master:
+            stack_versions['latest'] = {'beats': 'master', 'ecs': 'master'}
 
         # TODO: remove once py-eql supports ipv6 for cidrmatch
         # Or, unregister the cidrMatch function and replace it with one that doesn't validate against strict IPv4
         with eql.parser.elasticsearch_syntax, eql.parser.ignore_missing_functions:
             parsed = eql.parse_query(self.query)
 
         beat_types = [index.split("-")[0] for index in indexes if "beat-*" in index]
-        beat_schema = beats.get_schema_from_eql(parsed, beat_types, version=beats_version) if beat_types else None
 
-        for version in ecs_versions:
-            schema = ecs.get_kql_schema(indexes=indexes, beat_schema=beat_schema, version=version)
+        for stack_version, mapping in stack_versions.items():
+            beats_version = mapping['beats']
+            ecs_version = mapping['ecs']
+            err_trailer = f'stack: {stack_version}, beats: {beats_version}, ecs: {ecs_version}'
+
+            beat_schema = beats.get_schema_from_eql(parsed, beat_types, version=beats_version) if beat_types else None
+
+            schema = ecs.get_kql_schema(indexes=indexes, beat_schema=beat_schema, version=ecs_version)
             eql_schema = ecs.KqlSchema2Eql(schema)
 
             try:
                 # TODO: switch to custom cidrmatch that allows ipv6
                 with eql_schema, eql.parser.elasticsearch_syntax, eql.parser.ignore_missing_functions:
                     eql.parse_query(self.query)
-
-            except eql.EqlTypeMismatchError:
-                raise
-
             except eql.EqlParseError as exc:
                 message = exc.error_msg
-                trailer = None
+                trailer = err_trailer
                 if "Unknown field" in message and beat_types:
-                    trailer = "\nTry adding event.module or event.dataset to specify beats module"
+                    trailer = f"\nTry adding event.module or event.dataset to specify beats module\n\n{trailer}"
 
                 raise exc.__class__(exc.error_msg, exc.line, exc.column, exc.source,
                                     len(exc.caret.lstrip()), trailer=trailer) from None
+            except Exception as exc:
+                exc.args += (err_trailer,)
```

Comment:
hmm, I wonder if mutating an Exception is a good idea.
was this a code path you saw get hit and what were the exception types?

Reply:
I could raise a custom exception instead if that makes you more comfortable. Essentially, with iterative validation, it is a must to know which version the validation failed on or else really difficult to t/s

Reply:
df4ac2b